global_schema.ddl

```diff
@@ -0,0 +1,294 @@
+	agent_version	TEXT,
+	CONSTRAINT		fk_controller_node_controller
+		FOREIGN KEY	(controller_uuid)
+		REFERENCES	controller(uuid)
+);
+
+/*
+ * Upgrade Info
+ */
+CREATE TABLE IF NOT EXISTS upgrade_info (
+	uuid				TEXT PRIMARY KEY,
+	id					TEXT NOT NULL,
+	previous_version	TEXT,
+	target_version		TEXT,
+	status				TEXT,
+	started				TIMESTAMP
+);
+
+CREATE UNIQUE INDEX IF NOT EXISTS idx_upgrade_info_id
+ON upgrade_info (id);
+
+CREATE TABLE IF NOT EXISTS controller_node_upgrade_ready (
+	uuid					TEXT PRIMARY KEY,
+	controller_node_uuid	TEXT NOT NULL,
+	upgrade_info_id			TEXT NOT NULL,
+	CONSTRAINT				fk_upgrade_ready_controller
+		FOREIGN KEY			(controller_node_uuid)
+		REFERENCES			controller_node(uuid),
+	CONSTRAINT				fk_upgrade_ready_upgrade_info
+		FOREIGN KEY			(upgrade_info_id)
+		REFERENCES			upgrade_info(id)
+);
+
+CREATE UNIQUE INDEX IF NOT EXISTS idx_controller_node_upgrade_ready_cnid_uiid
+ON controller_node_upgrade_ready (controller_node_uuid, upgrade_info_id);
+
+CREATE TABLE IF NOT EXISTS controller_node_upgrade_done (
+	uuid					TEXT PRIMARY KEY,
+	controller_node_uuid	TEXT NOT NULL,
+	upgrade_info_id			TEXT NOT NULL,
+	CONSTRAINT				fk_upgrade_done_controller
+		FOREIGN KEY			(controller_node_uuid)
+		REFERENCES			controller_node(uuid),
+	CONSTRAINT				fk_upgrade_done_upgrade_info
+		FOREIGN KEY			(upgrade_info_id)
+		REFERENCES			upgrade_info(id)
+);
+
+CREATE UNIQUE INDEX IF NOT EXISTS idx_controller_node_upgrade_done_cnid_uiid
+ON controller_node_upgrade_done (controller_node_uuid, upgrade_info_id);
+
+/*
+ * Models
+ */
+CREATE TABLE IF NOT EXISTS model (
+	uuid					TEXT PRIMARY KEY,
+	name					TEXT,
+	type					TEXT,
+	life					INT,
+	controller_uuid			TEXT NOT NULL,
+	migration_mode			TEXT,
+	environ_version			INT,
+	cloud_uuid				TEXT NOT NULL,
+	cloud_region_uuid		TEXT,
+	latest_available_tools	TEXT,
+	password_hash			TEXT,
+	force_destroyed			BOOLEAN,
+	destroy_timeout			INT,
+	controller_model		BOOLEAN,
+	-- owner
+	-- cloud-credential
+	-- SLA
+	-- meter-status
+	CONSTRAINT				fk_model_controller
+		FOREIGN KEY			(controller_uuid)
+		REFERENCES			controller(uuid),
+	CONSTRAINT				fk_model_cloud
+		FOREIGN KEY			(cloud_uuid)
+		REFERENCES			cloud(uuid),
+	CONSTRAINT				fk_model_cloud_region
+		FOREIGN KEY			(cloud_region_uuid)
+		REFERENCES			cloud_region(uuid)
+);
+
+/*
+ * Users
+ */
+CREATE TABLE IF NOT EXISTS user (
+	uuid					TEXT PRIMARY KEY,
+	name					TEXT NOT NULL,
+	display_name			TEXT,
+	deactivated				BOOLEAN,
+	deleted					BOOLEAN,
+	secret_key				TEXT,
+	password_hash			TEXT,
+	password_salt			TEXT,
+	created_by_user_uuid	TEXT NOT NULL,
+	date_created			TIMESTAMP,
+	last_login				TIMESTAMP,
+	last_login_model_uuid	TEXT NOT NULL,
+	CONSTRAINT				fk_user_created_by_user
+		FOREIGN KEY			(created_by_user_uuid)
+		REFERENCES			user(uuid),
+	CONSTRAINT				fk_user_last_login_model
+		FOREIGN KEY			(last_login_model_uuid)
+		REFERENCES			model(uuid)
+);
+
+CREATE UNIQUE INDEX IF NOT EXISTS idx_user_name
```

Comment:
We don't want this constraint any longer - there's an open bug somewhere around the fact that we can't re-use a user name that has been deactivated in the past. We'll have to do app-level checking that disables active users with the same name.

Reply:
It can stay as an index, just not unique